File: apps/mobile/shared/Sources/FlagshipAPI/Client/FrontPageClient.swift
import Foundation

/// Box-direct delivery for the owner-assignable apex ("front page").
///
/// Two unauthenticated reads + one signature-authed write, all on the box's
/// pinned canonical pipe (same trust posture as LockPowerClient):
///
///   GET  https://<serverDomain>/api/front-page   → { label, active }
///   GET  https://<serverDomain>/api/services     → { apps: [{ urlLabel, name, … }] }
///   POST https://<serverDomain>/api/front-page   (set-front-page PhoneOrder envelope)
///
/// The transport takes an already-built `request` dictionary + the hex
/// signature so this layer stays free of the FlagshipCore envelope types
/// (FlagshipCore depends on FlagshipAPI, not the reverse).
public protocol FrontPageClient: Sendable {
    func getFrontPage(serverDomain: String) async throws -> FrontPageState
    func listFrontPageOptions(serverDomain: String) async throws -> [FrontPageOption]
    /// POST a signed `set-front-page` PhoneOrder. Throws on non-2xx.
    func setFrontPage(serverDomain: String, request: [String: Any], signatureHex: String) async throws
}

public struct FrontPageState: Equatable, Sendable {
    /// Assigned service url-label; nil = default Flagship page.
    public let label: String?
    /// Whether the assigned label currently resolves to an installed service.
    public let active: Bool
    public init(label: String?, active: Bool) {
        self.label = label
        self.active = active
    }
}

public struct FrontPageOption: Equatable, Sendable, Identifiable {
    public let urlLabel: String
    public let name: String
    public var id: String { urlLabel }
    public init(urlLabel: String, name: String) {
        self.urlLabel = urlLabel
        self.name = name
    }
}

/// URLSession-backed implementation. Rides the BOX-pinned session (hard-fail
/// cert-fingerprint pinning) exactly like the screens client.
public final class LiveFrontPageClient: FrontPageClient, @unchecked Sendable {
    private let urlSession: URLSession

    public init(urlSession: URLSession) {
        self.urlSession = urlSession
    }

    private static func baseUrl(_ serverDomain: String) -> String {
        let host = serverDomain.trimmingCharacters(in: CharacterSet(charactersIn: "/ "))
        return "https://\(host)"
    }

    private func send(_ req: URLRequest) async throws -> Data {
        let data: Data
        let resp: URLResponse
        do {
            (data, resp) = try await urlSession.data(for: req)
        } catch {
            if let host = req.url?.host,
               CertPinMismatchSink.shared.consumeRecentMismatch(host: host) {
                throw ScreensClientError.certPinMismatch(host: host)
            }
            throw error
        }
        guard let http = resp as? HTTPURLResponse else {
            throw ScreensClientError.http(status: 0, message: "no response")
        }
        if !(200..<300).contains(http.statusCode) {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw ScreensClientError.http(status: http.statusCode, message: text)
        }
        return data
    }

    private func get(serverDomain: String, path: String) async throws -> Data {
        guard let url = URL(string: Self.baseUrl(serverDomain) + path) else {
            throw ScreensClientError.http(status: 0, message: "bad URL")
        }
        return try await send(URLRequest(url: url))
    }

    public func getFrontPage(serverDomain: String) async throws -> FrontPageState {
        let data = try await get(serverDomain: serverDomain, path: "/api/front-page")
        let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        let label = obj?["label"] as? String
        let active = (obj?["active"] as? Bool) ?? false
        return FrontPageState(label: label, active: active)
    }

    public func listFrontPageOptions(serverDomain: String) async throws -> [FrontPageOption] {
        let data = try await get(serverDomain: serverDomain, path: "/api/services")
        let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        let apps = (obj?["apps"] as? [[String: Any]]) ?? []
        return apps.compactMap { a in
            guard let label = a["urlLabel"] as? String else { return nil }
            return FrontPageOption(urlLabel: label, name: (a["name"] as? String) ?? label)
        }
    }

    public func setFrontPage(serverDomain: String, request: [String: Any], signatureHex: String) async throws {
        guard let url = URL(string: Self.baseUrl(serverDomain) + "/api/front-page") else {
            throw ScreensClientError.http(status: 0, message: "bad URL")
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "content-type")
        req.httpBody = try JSONSerialization.data(
            withJSONObject: ["request": request, "signature": signatureHex], options: [])
        _ = try await send(req)
    }
}

/// In-memory mock: records what was set (so tests can assert wire shapes)
/// and returns configurable state + options.
public final class MockFrontPageClient: FrontPageClient, @unchecked Sendable {
    public struct Sent: Sendable {
        public let serverDomain: String
        public let request: [String: String]
        public let signatureHex: String
    }

    private let lock = NSLock()
    private var _sent: [Sent] = []
    public var sent: [Sent] { lock.withLock { _sent } }

    public var state = FrontPageState(label: nil, active: false)
    public var options: [FrontPageOption] = []
    /// Optional error to throw on the next call (any path), then cleared.
    public var nextError: Error?

    public init() {}

    private func maybeThrow() throws {
        if let e = nextError { nextError = nil; throw e }
    }

    public func getFrontPage(serverDomain: String) async throws -> FrontPageState {
        try maybeThrow()
        return state
    }

    public func listFrontPageOptions(serverDomain: String) async throws -> [FrontPageOption] {
        try maybeThrow()
        return options
    }

    public func setFrontPage(serverDomain: String, request: [String: Any], signatureHex: String) async throws {
        try maybeThrow()
        var flat: [String: String] = [:]
        for (k, v) in request { flat[k] = String(describing: v) }
        let label = flat["label"] ?? ""
        lock.withLock { _sent.append(Sent(serverDomain: serverDomain, request: flat, signatureHex: signatureHex)) }
        state = FrontPageState(label: label.isEmpty ? nil : label, active: !label.isEmpty)
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/api/FrontPageClient.kt
// Pod-direct client for the owner-assignable apex ("front page").
//
// Two unauthenticated reads + one owner-IRK-signed write, straight to the
// box (`https://<pod>`), mirroring LockPowerClient + the daemon shapes in
// packages/server-daemon/src/frontPage.ts:
//
//   GET  https://<pod>/api/front-page   → { label, active }
//   GET  https://<pod>/api/services     → { apps: [{ urlLabel, name, … }] }
//   POST https://<pod>/api/front-page   { request, signature }  (IRK)

package com.flagshipserver.app.api

import com.flagshipserver.app.core.JsonHttpTransport
import com.flagshipserver.app.core.OkHttpJsonTransport
import kotlinx.serialization.Serializable

@Serializable
data class FrontPageState(
    /** Assigned service url-label; null = default Flagship page. */
    val label: String? = null,
    /** Whether the assigned label currently resolves to an installed service. */
    val active: Boolean = false,
)

@Serializable
data class FrontPageServiceEntry(
    val urlLabel: String,
    val name: String? = null,
)

@Serializable
data class FrontPageServicesResponse(
    val apps: List<FrontPageServiceEntry> = emptyList(),
)

@Serializable
data class SetFrontPageInner(
    val type: String = "set-front-page",
    val serverId: String,
    /** Service url-label to front-page; "" clears (default page). */
    val label: String,
    val issuedAt: Long,
)

@Serializable
data class SetFrontPageRequest(
    val request: SetFrontPageInner,
    /** Hex Ed25519 signature over SetFrontPageOrder.canonicalBytes. */
    val signature: String,
)

/** Daemon `/api/front-page` ack — `{ ok, label }`. */
@Serializable
data class FrontPageAck(val ok: Boolean = false, val label: String? = null)

/** Pod-direct client. [podBaseUrl] resolves `https://<serverDomain>` (the
 *  FQDN IS the pod). Pluggable transport for tests. */
class FrontPageClient(
    private val transport: JsonHttpTransport = OkHttpJsonTransport(),
    private val podBaseUrl: (serverDomain: String) -> String = { "https://$it" },
) {
    suspend fun getFrontPage(serverDomain: String): FrontPageState =
        transport.getJson(
            "${podBaseUrl(serverDomain)}/api/front-page",
            FrontPageState.serializer(),
        )

    suspend fun listOptions(serverDomain: String): List<FrontPageServiceEntry> =
        transport.getJson(
            "${podBaseUrl(serverDomain)}/api/services",
            FrontPageServicesResponse.serializer(),
        ).apps

    suspend fun setFrontPage(serverDomain: String, body: SetFrontPageRequest): FrontPageAck =
        transport.postJsonForResponse(
            "${podBaseUrl(serverDomain)}/api/front-page",
            body,
            SetFrontPageRequest.serializer(),
            FrontPageAck.serializer(),
        )
}
